File: src/main/kotlin/plodsoft/mygvm/util/ColorUtil.java
package plodsoft.mygvm.util;

public class ColorUtil {
    /**
     * 解析#rrggbb或#rgb格式的颜色字符串
     * @param color
     * @return 0xrrggbb
     * @throws IllegalArgumentException
     */
    public static int parse(String color) {
        if (!color.matches("^#([a-fA-F0-9]{3}|[a-fA-F0-9]{6})$")) {
            throw new IllegalArgumentException("Invalid color string");
        }

        if (color.length() == 4) {
            int r = Character.digit(color.charAt(1), 16);
            int g = Character.digit(color.charAt(2), 16);
            int b = Character.digit(color.charAt(3), 16);
            return r << 20 | r << 16 | g << 12 | g << 8 | b << 4 | b;
        } else {
            int r = Character.digit(color.charAt(1), 16) << 4 | Character.digit(color.charAt(2), 16);
            int g = Character.digit(color.charAt(3), 16) << 4 | Character.digit(color.charAt(4), 16);
            int b = Character.digit(color.charAt(5), 16) << 4 | Character.digit(color.charAt(6), 16);
            return r << 16 | g << 8 | b;
        }
    }
}

File: src/main/kotlin/plodsoft/mygvm/util/ColorUtil.kt
package plodsoft.mygvm.util

/**
 * 解析#rrggbb或#rgb格式的颜色字符串
 *
 * @return 0xrrggbb
 *
 * @throws IllegalArgumentException
 */
fun String.parseColor(): Int {
    if (!matches("^#([a-fA-F0-9]{3}|[a-fA-F0-9]{6})$".toRegex())) {
        throw IllegalArgumentException("Invalid color string")
    }

    return if (length == 4) {
        val r = Character.digit(this[1], 16)
        val g = Character.digit(this[2], 16)
        val b = Character.digit(this[3], 16)
        (r shl 20) or (r shl 16) or (g shl 12) or (g shl 8) or (b shl 4) or b
    } else {
        val r = Character.digit(this[1], 16) shl 4 or Character.digit(this[2], 16)
        val g = Character.digit(this[3], 16) shl 4 or Character.digit(this[4], 16)
        val b = Character.digit(this[5], 16) shl 4 or Character.digit(this[6], 16)
        (r shl 16) or (g shl 8) or b
    }
}
